srk/src/transition.ml

```diff
@@ -390,63 +437,142 @@ struct
+    Smt.Solver.reset solver;
+    let trs = List.map rename_skolems trs in 
+    interpolate_query ~solver:solver trs post (fun _ _ _ -> `Invalid) @@ interpolate_unsat_core qflia_solver
+
+  let interpolate_or_concrete_model ?(solver=Smt.mk_solver C.context) ?(qflia_solver=(Smt.mk_solver ~theory:"QF_LIA" srk)) trs post = 
+    (* subst_model: rename skolem constants back to their appropriate names using reverse subscript table *)
+    Smt.Solver.reset solver;
+    let trs = List.map rename_skolems trs in 
+    let sat_model model (symbols: Symbol.Set.t list) ss_inv = 
+        let m = 
+          List.fold_left (fun m' symbols -> 
+            Symbol.Set.fold (fun s m -> 
+              (* the provided model is over both subscripted vocabulary and original vocabulary *)
+              begin match Hashtbl.find_opt ss_inv s with 
+              | Some s' -> (* subscripted variable *)
+                Interpretation.add s' (Interpretation.value model s) m
+              |  None -> (* non-subscripted; query directly *)
+                Interpretation.add s (Interpretation.value model s)  m 
+              end) symbols m'
+          ) (Interpretation.empty srk) symbols in 
+        (* symbols is a list of subscripted symbols arranged in left-to-right order. 
+          folding over this in left-to-right order amounts to forward concrete execution. *)
+        `Invalid m
+    in interpolate_query ~solver:solver trs post sat_model @@ interpolate_unsat_core qflia_solver
+
+
+  let extrapolate ?(solver=Smt.mk_solver srk) t1 t2 t3 : [`Sat of (C.t formula * C.t formula) | `Unsat ] =
+  (* Create fresh copies of skolem variables: nondet consts, loop tripcount variables, etc. *)
+  let t1, t2, t3 = rename_skolems t1, rename_skolems t2, rename_skolems t3 in 
+  (* Perform variable subscripting; equivalent to "SSA"ing the formulas t1, t2, t3 *)
+  let subscript_tbl, reverse_subscript_tbl = Hashtbl.create 991, Hashtbl.create 991 in
+  let subscript sym =
+    try
+      Hashtbl.find subscript_tbl sym
+    with Not_found -> mk_const srk sym
+  in
+  (* Similar as interpolation (but without the indicators part): 
+     Convert tr into a formula, and simultaneously update the subscript table *)
+  let to_ss_formula tr =
+    let ss_guard = substitute_const srk subscript (guard tr)
+    in let (ss, phis) =
+      M.fold (fun var term (ss, phis) ->
+          let var_sym = Var.symbol_of var in
+          let var_ss_sym = mk_symbol srk (Var.typ var :> typ) in
+          let var_ss_term = mk_const srk var_ss_sym in
+          let term_ss = substitute_const srk subscript term in
+          ((var_sym, var_ss_term, var_ss_sym)::ss,
+            (mk_eq srk var_ss_term term_ss)::phis))
+        tr.transform
+        ([], [ ss_guard ])
+    in
+    List.iter (fun (k, v, l) -> Hashtbl.add subscript_tbl k v; Hashtbl.add reverse_subscript_tbl l k) ss;
+    mk_and srk phis
+  in let ss_t1, ss_t2, ss_t3 = to_ss_formula t1, to_ss_formula t2, to_ss_formula t3
```

Comment:
Nitpick: this "in let" style is unsual -- I'd prefer
  `let conj = mk_and srk [ss_t1; ss_t2; ss_t3] in`
over
  `in let conj = mk_and srk [ss_t1; ss_t2; ss_t3]`